cluster/kubernetes/sync.go

```diff
@@ -2,31 +2,308 @@ package kubernetes
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+const (
+	syncSetLabel       = kresource.PolicyPrefix + "sync-set"
+	checksumAnnotation = kresource.PolicyPrefix + "sync-checksum"
+
+	// The namespace to presume if something doesn't have one, and we
+	// haven't been told what to use as a fallback. This is what
+	// `kubectl` uses when there's no config setting the fallback
+	// namespace.
+	DefaultFallbackNamespace = "default"
 )
 
+// Sync takes a definition of what should be running in the cluster,
+// and attempts to make the cluster conform. An error return does not
+// necessarily indicate complete failure; some resources may succeed
+// in being synced, and some may fail (for example, they may be
+// malformed).
+func (c *Cluster) Sync(spec cluster.SyncSet) error {
+	logger := log.With(c.logger, "method", "Sync")
+
+	// Keep track of the checksum of each resource, so we can compare
+	// them during garbage collection.
+	checksums := map[string]string{}
+
+	// NB we get all resources, since we care about leaving unsynced,
+	// _ignored_ resources alone.
+	clusterResources, err := c.getResourcesBySelector("")
+	if err != nil {
+		return errors.Wrap(err, "collating resources in cluster for sync")
+	}
+
+	cs := makeChangeSet()
+	var errs cluster.SyncError
+	for _, res := range spec.Resources {
+		id := res.ResourceID().String()
+		// make a record of the checksum, whether we stage it to
+		// be applied or not, so that we don't delete it later.
+		csum := sha1.Sum(res.Bytes())
+		checkHex := hex.EncodeToString(csum[:])
+		checksums[id] = checkHex
+		if res.Policy().Has(policy.Ignore) {
+			logger.Log("info", "not applying resource; ignore annotation in file", "resource", res.ResourceID(), "source", res.Source())
+			continue
+		}
+		if cres, ok := clusterResources[id]; ok && cres.Policy().Has(policy.Ignore) {
+			logger.Log("info", "not applying resource; ignore annotation in cluster resource", "resource", cres.ResourceID())
+			continue
+		}
+		resBytes, err := applyMetadata(res, spec.Name, checkHex)
+		if err == nil {
+			cs.stage("apply", res.ResourceID(), res.Source(), resBytes)
+		} else {
+			errs = append(errs, cluster.ResourceError{ResourceID: res.ResourceID(), Source: res.Source(), Error: err})
+			break
+		}
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.muSyncErrors.RLock()
+	if applyErrs := c.applier.apply(logger, cs, c.syncErrors); len(applyErrs) > 0 {
+		errs = append(errs, applyErrs...)
+	}
+	c.muSyncErrors.RUnlock()
+
+	if c.GC {
+		orphanedResources := makeChangeSet()
+
+		clusterResources, err := c.getResourcesInSyncSet(spec.Name)
+		if err != nil {
+			return errors.Wrap(err, "collating resources in cluster for calculating garbage collection")
+		}
+
+		for resourceID, res := range clusterResources {
+			actual := res.GetChecksum()
+			expected, ok := checksums[resourceID]
+
+			switch {
+			case !ok: // was not recorded as having been staged for application
```

Comment:
As I understand it, this will cause the deletion of resources whose manifest has been made invalid (potentially by mistake) and thus, not included in the syncset. This means that:

1. corrupting a manifest file by mistake is equivalent to deleting it
2. renaming the extension of a file by mistake is equivalent to deleting it

I am not sure we can/should do anything about 2. But I think we should have a mechanism in place for (1)

Reply:
There's a few ways a file can be invalid:

 - no longer recognised as being a YAML file: I agree, nothing much we can do about this
 - if it's invalid YAML: we bail on parsing if any files are invalid YAML, so we wouldn't get this far
 - if it's valid YAML, but not accepted by kubectl apply: we still explicitly record these to prevent them being garbage-collected

Reply:
Oh, perfect then!